ahds/ahds_common.py

```diff
@@ -400,8 +197,28 @@ def move_attr(self,to,name):
         setattr(self,to,_tomove)
         delattr(self,name)
 
+    def __setattr__(self,name,value):
```

Comment:
It's very hard to follow what's going on here. Is there another way to do this? I would suggest avoiding exotic interpreter objects.

Reply:
I'd which there would be. What i have found is:

If sticking to attributes only which reduces headache of mixing properties with load on demmand semantics of attributes (dynamic properties are even bigger headache as properties are always to be defined on class and not instance) i haven't yet found any simpler and less magic approach to prevent attribute assignment through setattr and assignment operator = from outside class and subclass scope.

The only alternative would be a hidden `_lock` attribute which causes `__setattr__` to throw `AttributeError` whenever `_lock` is set to `True` and only accept assignment if set to `False`. But that would require to explicitly set `_lock` to `False` in in `add_attr` and any other methods of Block class and subclasses which wants to directly assining value to any attribute  directly including all`__init__` methods along the class hierarchie. And properly reset to previous state found on entering. Which is more error prone than having `Block.__setattr__` check the scope from within which the assignment is attempted and grant it if it comming from within Block class or subclass method.

Another approach i have neither tested nor tried and have no idea whether this is worth the efforts is by defining via custom meta class for Block. I wouldn't suggest this route for now as this adds complexity of defining Block meta class to code. Which i currently do not see any much benefit for the efforts needed adding even more magic than by simply inspecting callstack frame.

The only thing to reduce magic a bit could be to replace a variable by _calling_context_stack_frame and add comments which are anyway missing throughout more places of the code. Possibly also use methods from official python `inspect` module instead of calling hidden `sys._getframe` method but would not change anything in the logic.

Reply:
I think i have found a good compromise. Any dynamic attributes defined by AmriaMesh and HyperSurface files which are stored inside __dict__ can only be added or modified though a call to add_addr and renamed by calling move_attr. Any other attribute has to be declared thorugh __slots__ and may be modified by direct assignment and through setattr calls.
Some protection is provided by pythons protected and private attribute mimickry (prepending by one ore two '_'). They are not really read only but by convention nobody will modify them from outside class without very good reason. In addintion attributes starting with two __ are rewritten to _<ClassName><attributename> Which is quite annoying when trying to display their content in debugsessions or apply filter operations for example in __getattribute__ to them. Which is why i mean while stick to protected attributes and method names only where possible.
Any public attribute has to be explicitly write protected by explicitly overloading __setattr__ method and comparing attribute name against list of read only attribute names and if found throwing an Attribute error while other wise passing on request.

Reply:
I like what I see. It's far more readable. Let me go through in detail then I'll get back to you.

Reply:
I just noticed, the new solution has a slight bug on line 203. It allows setting of attribute even if not in slots of `self.__class__` which is the first along `__mro__` to be checked.  The whole block should read:

```python
for _found in (
    True
    for _slots in (
        getattr(_parent,"__slots__",None)
        for _parent in super(Block,self).__getattribute__("__class__").__mro__
    )
    if _slots is not None and name in _slots
):
    return True
```

This will be fixed in the next commit to assure all `__slots__` structures along `__mro__` of `self.__class__` are checked whether they contain the attribute or not.